### ISSUE TITLE
Inefficient Tile Ripping

### ISSUE THREAD

**Issue by gokiburikin** [YukkiTimmy/Tilemap2Tileset#3]
The actual tile ripping process is very slow compared to what it could be. As a proof of concept I've written a faster method, though it doesn't have the same amount of options. The current build of Tilemap2Tileset takes about 19 seconds for the Koholint map and this function averages about 0.7 seconds.

https://user-images.githubusercontent.com/3820082/142560827-b2785722-468d-4095-844a-9946a36e6a6c.mp4

This code is CC0 so feel free to do whatever you want with it, no credit needed. I'm sure it can also be improved even before simply rewriting it as a C++ module. Keep in mind this DOES utilize the `hash()` function so I believe in theory it could have collisions, but writing a quick test for collisions to double check their actual pixel data would be pretty easy.

```gdscript
func rip_unique_tiles( image:Image, cell_width:int = 8, cell_height:int = 8 ):
	var columns:int = int(floor( image.get_width() / cell_width ))
	var rows:int = int(floor( image.get_height() / cell_height ))
	var unique_cells := {}
	var image_data := image.get_data()
	var pixels = image.get_width() * image.get_height()
	var pixel_data_size:int = image_data.size() / pixels
	var cell_column_step := cell_width * pixel_data_size
	var cell_row_step := cell_column_step * columns
	for y in range( rows ):
		for x in range( columns ):
			var cell_data = []
			var cell_index = x * cell_column_step + y * cell_row_step * cell_height
			for i in range( cell_height ):
				var c = cell_index + cell_row_step * i
				cell_data.append_array( image_data.subarray( c, c + cell_column_step - 1 ) )
			var cell_hash = hash( cell_data )
			if not unique_cells.has( cell_hash ):
				unique_cells[cell_hash] = Vector2(x,y)
	
	var unique_tiles = Image.new()
	var minimized_texture_size = nearest_po2( unique_cells.size() )
	var minimized_columns = minimized_texture_size / cell_width
	var minimized_rows = minimized_texture_size / cell_height
	unique_tiles.create( minimized_texture_size, minimized_texture_size, false, image.get_format() )
	var unique_index := 0
	for entry in unique_cells.values():
		var cell_x = ( unique_index ) % minimized_columns
		var cell_y = floor( unique_index / minimized_columns )
		unique_tiles.blit_rect( image, Rect2( entry.x * cell_width, entry.y * cell_height, cell_width, cell_height ), Vector2( cell_x * cell_width, cell_y * cell_height ) )
		unique_index += 1
	var texture = ImageTexture.new()
	texture.create_from_image( unique_tiles )
	texture.flags ^= texture.FLAG_FILTER
	unique_tiles_image.texture = texture
	return unique_cells
```

**Comment by YukkiTimmy** [comment]
This is the smartets thing I have seen in a while ngl.
Someone already suggested using a hash function, but honestly I am just to dumb to use it properly :D

I will try to do my best implementing this! 

Thanks alot!

**Comment by AsherGlick** [comment]
#4 Implements hashing and handles rotated and flipped tiles.

**Comment by gokiburikin** [comment]
Based on the results of #4 I went back to check the speed of hashing the image data directly and it's indeed another order of magnitude faster!

https://user-images.githubusercontent.com/3820082/142649648-56932795-8c42-4931-aee9-213c3d784ab4.mp4

For completeness' sake I'll update this with an example of the yet much faster results. The main thing to look out for in #4 is the creation of new Image's constantly. It shouldn't be necessary since you can reuse the previously created Image and just blit over it again.

Unfortunately, the fastest way I could think to support rotations was too involved for a single function, so I've attached a minimal example project. This utilizes a Viewport that rotates the input image and uses that rotated viewport texture to grab a rotated copy of the texture to perform hashes on.

Edit: You could avoid this viewport shenaniganry by using the rotate image function found in #4. You'd ideally want to keep a copy of the rotated image around when the option to deduplicate based on orientation is enabled so that in the event you want to re-cut the tiles you don't have to rebuild the rotated image. It will be slower, but in my testing the viewport method is not perfect (rotating the images pixel by pixel results in 308 unique tiles, viewport results in 310.)

[tile-ripper.zip](https://github.com/YukkiTimmy/Tilemap2Tileset/files/7571303/tile-ripper.zip)

Like before, feel free to use anything in this project, CC0.

**Comment by YukkiTimmy** [comment]
Thanks alot to both of you for helping me out that much!
I implemented the hash function of AsherGlick and changed a few things here and there.
It works like a charm now and I had to change the second timer into a millisecond one :D

I  credited both of you in the source code and I will make another reddit post tomorrow where I credit you as well , I hope you are both fine with that.
Thanks so much again!
